### DISCUSSION TITLE
how do I call a public interface method from the ui?

### DISCUSSION THREAD

**Discussion by dckc** [Agoric/agoric-sdk#10372] [Q&A] [answered]
_I saw this question today in discord..._

Hi, how do I call a public interface method from the ui?

I can connect the vstorage, get the contract, use ChainStorageWatcher to get the instance but I don't find any tutorial on how to call a method defined in the public interface

_stay tuned for an answer. cc @Jovonni @amessbee_

**Comment by dckc** [comment]
As explained in [Building Client Dapps](https://docs.agoric.com/guides/getting-started/contract-rpc.html)

> Clients interact with the consensus layer by making queries and submitting messages in signed transactions.

If the public facet method is a "get", the client often wants not just one answer, but a stream of answers as the current value changes.

In that case, you can somewhat simulate a call on the public facet by
 - providing a notifier from the client's public interface
 - subscribing to the notifier from within the contract and for each update, feed the contents thru vstorage
 - querying / watching vstorage from a client

If the public facet method is not a "get", then you should be able to accomplish it with invitations and offers as discussed in the docs above.

_I can imagine this is overly brief. Please ask questions where it's not clear. Or perhaps one of my colleagues will find a moment to give examples._

**Reply by yribero** [reply]
My current intent is to create and deploy a contract for storing notes (simple strings) into an array on the contract and retrieve them with a getter, watching if new ones are added.
I defined two functions in my publicFacet: getNotes and addNote. I would like to call those methods, and I hope only addNote requires to create a transaction.
I can't find any examples how to achieve this result. Though I understand this explanation in principle, it is still way too abstract to turn it into working code. I appreciate any help. In the discord #dev channel I go by the name fenner.

**Reply by dckc** [reply, marked as answer]
The agoric platform doesn't have a special case for query methods in contracts. (_Whether it could or should is a longer discussion..._)

Instead, it has vstorage / chainStorage.

The addNote method can publish to chainStorage, and the ui code can query vstorage.

I see @amessbee put together a quick example where the contract does...

```js
      await storePatientData(patientData.patientId, patientData);
```
 -- https://github.com/amessbee/dapp-med-rep/blob/b0f3adb041d9383a7badee12b51bf457cce0e473/contract/src/med-rec-contract.js#L84C1-L84C66

The `publishHandler` function is close to an `addNote` function. Since this contract isn't actually trading assets, the indirection around `makePublishInvitation` is unfortunate overhead. We have sketches for avoiding that eventually, but for now, that's the way it works.

On the ui side, `getNotes` is implemented as a `fetch`:

```js
  const fetchPatientData = async (patientId: string) => {
    setSelectedPatientId(patientId); // Set the selected patient ID
    const response = await fetch(`${ENDPOINTS.API}/agoric/vstorage/data/published.patientData.patients.${patientId}`);
```
https://github.com/amessbee/dapp-med-rep/blob/b0f3adb041d9383a7badee12b51bf457cce0e473/ui/src/App.tsx#L333C1-L335C119

**Reply by yribero** [reply]
Hi, thank you very much for the pointer.

**Comment by dckc** [comment]
## EVM analog: views

- [Solidity \- View and Pure Functions \- GeeksforGeeks](https://www.geeksforgeeks.org/solidity-view-and-pure-functions/)

**Reply by yribero** [reply]
I rewrote my contract and could create a note:
![published-note](https://github.com/user-attachments/assets/6b336679-02bd-4ef7-8479-c258fb8b68c6)
I can see a note object appearing on the vstorage:
![vstorage](https://github.com/user-attachments/assets/86bd67b4-74b7-4aba-9963-da6a375432ff)
But I currently can't navigate the vstorage and see all the notes (nothing happens when I click on the inner note object, I need to access them one by one using the API shown in picture 1)

How can I react (as a neutral observer) to the publishing of notes? I would like to show them on App.tsx as they appear in real time

**Reply by dckc** [reply]
> How can I react (as a neutral observer) to the publishing of notes?

The [chainStorageWatcher](https://docs.agoric.com/guides/getting-started/ui-tutorial/querying-vstorage.html#chainstoragewatcher) API is designed for this. It's basically HTTP polling.

In order to react to new notes, you might want a `published.notes.latest` that increments each time a new note becomes available.

**Reply by yribero** [reply]
I managed to access the vstorage from my web app and read all the nodes I was interested in. Now I would like to do the same within the smart contract, but this seems trickier.
I have found a way to access the agoricNames

`const ns = (await makeAgoricNamesAccess()).agoricNames`

But that's not what I want: I need to access the notes (which are not under agoric names) but I am currently not able to. Any hints?

**Reply by yribero** [reply]
Some more context from discord:

How do I access nodes create d in a smart contract? For example I can create a collection in my contract like this:

`const notes = await E(privateArgs.storageNode).makeChildNode('collection');`

And then push a note into it:

```
    const noteNode = await E(notes).makeChildNode(note.id.toString());
    const marshalledData = JSON.stringify(await E(marshaller).toCapData(note));
    await E(noteNode).setValue(marshalledData);
```

But later I'm not able to read them. These nodes don't have a getValue method, or any value/children reading method, for that matters.

I can access agoricNames and lookup stuff

`const ns = (await makeAgoricNamesAccess()).agoricNames`

But not my data whose root is at the same level as the agoricNames

**Reply by dckc** [reply]
> How do I access nodes create d in a smart contract?

chainStorage, the contract-side API, is write-only.
vstorage, the client-side API, is read-only.

As I mentioned [Oct 30](#discussioncomment-11103094), the norm is:
  - a contract provides a public facet method such as `getPublicTopics()`. Peers inside the VM can use that to subscribe to updates.
    - in this case, a `latestNoteId` topic would let peers inside the VM subscribe to new notes
    - an `E(publicFacet).getNote(id)` method would let peers inside the VM get the contents of any note
  - the contract subscribes to its own topics and reflects them into vstorage for off-chain clients

@amessbee perhaps you could update your demo app to show this pattern?

@yribero I can elaborate during [office hours](https://github.com/Agoric/agoric-sdk/wiki/Office-Hours). The next one is tomorrow.

**Reply by yribero** [reply]
Thank you once more. I will try to attend the event
